### PR TITLE
Feat/individual optimistic oracles

### PR DESCRIPTION
**Pull request by amarcu**
Refactored the current monolithic **OptimisticOracle** with multiple Validators into multiple specialized OptimisticOracles that manage specific assets. 

Instead of unique validators now we extend the abstract OptimisticOracle and implement specialized oracles. Therefore the old **ChainlinkValidator** is now  **OptimisticChainlinkOracle**, it still has the same `validate()` and `value()` functions and we added an `encodeNonce()` and `decodeNonce()` that handles the new nonce format.

The flow is the same, the most significant difference is how the `nonce` is computed and used:
- When a proposal is made by calling `shift()`, a new `nonce` is generated that contains the block timestamp at which the proposal was made. The `proposeTimestamp` is then used to validate the **propose** and the **dispute** window. 
- The shift method now requires `bytes` as `data` (instead of the bytes32 generated nonce) which is used in the nonce generation. This helps us in keeping a common `shift()` / `dispute()` interface for the Chainlink Oracle and any Proof Oracles that we would implement.
- The **OptimisticChainlinkOracle** `validate()` uses `proposeTimestamp` to check if the chainlink feed round data is stale or not.
- `_push()` will no longer trigger reverts if it fails when new values are proposed via `shift()`
- `bond()` is now guarded and requires proposers to be whitelisted.
- Added a simple guarded `lock()` mechanism that unregisters multiple `rateIds`(all that are managed by the oracle) and thus blocks further price updates to be pushed to **Collybus**. 
- Added a `withdraw()` method that allows proposers to unbond in case a rate was unregistered from the oracle.

The **OptimisticChainlinkOracle** also implements the `_push()` method that calls into **Collybus**.

We removed the rateConfig struct because we do not need to store a validator address anymore and we simplified `registerRateConfig()` and `unregisterRateConfig()`

### CODE REVIEW

File: .solhint.json [hunk 1]
```diff
@@ -13,6 +13,7 @@
     "reason-string": ["warn", { "maxLength": 64 }],
     "avoid-low-level-calls": "off",
     "no-inline-assembly": "off",
+    "no-empty-blocks": "off",
```
Comment (SolcAndMe) [inline .solhint.json:16]: remove

Comment (amarcu) [reply .solhint.json:16]: If we remove this line we will get linter warnings for this block:
```
        try
            ICollybus(collybus).updateSpot(
                address(uint160(uint256(rateId))),
                value_
            )
        {} catch {}
```
And adding ``// solhint-disable-next-line no-empty-blocks`` for both blocks doesn't look pretty either.

File: src/interfaces/IOptimisticOracle.sol [hunk 2]
```diff
@@ -0,0 +1,77 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+pragma solidity ^0.8.0;
+
+interface IOptimisticOracle {
+    /// @notice Derive the nonce of a proposal from `data` and other protocol specific parameters.
+    /// @dev Must revert if the `disputeWindow` is still active
+    /// @param prevNonce Nonce of the previous proposal
+    /// @param data Data of the current proposal
+    /// @return nonce of the current proposal
+    function encodeNonce(bytes32 prevNonce, bytes memory data)
```
Comment (SolcAndMe) [inline src/interfaces/IOptimisticOracle.sol:10]: add decodeNonce method

File: src/OptimisticOracle.sol [hunk 3]
```diff
@@ -130,6 +130,9 @@ abstract contract OptimisticOracle is IOptimisticOracle, Guarded {
             revert OptimisticOracle__unsetRateConfig_noRateConfig(rateId);
         }
 
+        // Clear the current proposal if it exists
```
Comment (SolcAndMe) [inline src/OptimisticOracle.sol:133]: describe why we need to clear it

Comment (amarcu) [reply src/OptimisticOracle.sol]: Updated the comment.

**Pull request closed without merging**